File: compiler/server/documentation.go
package server

import (
	"fmt"
	"wipple/database"
	"wipple/feedback"
	"wipple/queries"
)

type DocumentationRequest struct {
	InputMetadata
	Name string `json:"name"`
}

type DocumentationResponse struct {
	Documentation *Documentation `json:"documentation,omitempty"`
}

type Documentation struct {
	Declaration string `json:"declaration"`
	Docs        string `json:"docs"`
}

func (request *DocumentationRequest) handle() (DocumentationResponse, error) {
	libraryName := request.InputMetadata.Library
	if libraryName == "" {
		return DocumentationResponse{}, fmt.Errorf("missing library")
	}

	db, _, err := compileLibrary(libraryName)
	if err != nil {
		return DocumentationResponse{}, err
	}

	filter := func(node database.Node) bool {
		return true
	}

	data, ok := database.ContainsNode(db, func(node database.Node) (queries.DocumentationData, bool) {
		var data queries.DocumentationData
		ok := false
		queries.Documentation(db, node, filter, func(name string, d queries.DocumentationData) {
			if name == request.Name {
				ok = true
				data = d
			}
		})

		return data, ok
	})

	if !ok {
		return DocumentationResponse{}, nil
	}

	render := feedback.NewRender(db)
	render.WriteComments(data.Comments)
	docs := render.Finish()

	return DocumentationResponse{
		Documentation: &Documentation{
			Declaration: data.Declaration,
			Docs:        docs,
		},
	}, nil
}

File: compiler/server/documentation.rs
use crate::{InputMetadata, compile_library};
use serde::Deserialize;
use serde_json::json;
use std::collections::HashMap;
use wipple::{feedback::FeedbackWriter, queries};

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Request {
    #[serde(flatten)]
    metadata: InputMetadata,
}

pub async fn handle(request: Request) -> anyhow::Result<serde_json::Value> {
    let library_name = request
        .metadata
        .library
        .ok_or_else(|| anyhow::format_err!("missing library"))?;

    let (db, _) = compile_library(&library_name).await?;

    let mut items = HashMap::new();
    for data in queries::all(&db, queries::documentation) {
        let Some(name) = data.name else {
            continue;
        };

        let mut docs = String::new();
        FeedbackWriter::new(&db, &mut docs).write_comments(&data.comments);

        items.insert(
            name,
            json!({
                "declaration": data.declaration,
                "kind": data.kind,
                "docs": docs,
            }),
        );
    }

    Ok(json!({ "items": items }))
}
